### PR TITLE
add nix-store --query --valid-derivers command

### PR DESCRIPTION
**Pull request by symphorien**
# Motivation
On typical nixos installations, nixos-rebuild creates .drv files for all store paths of the system, but the deriver field of the db comes from hydra. If a dependency of a fixed output derivation changed between the time hydra built something and nixos-rebuild was called, the local .drv and the deriver provided by hydra will not match. In this case,
nix-store --query --deriver /store/path returns a missing .drv file, whereas we have at least one other deriver of this store path locally.

This commit adds a new command `nix-store --query --valid-derivers` which prints .drv files which can build the specified output *and actually exist locally*.
The intended final use case of this contribution is to be able to link an executable in the store to the store path of its source via the src field of its deriver. (specifically: nixseparatedebuginfod uses that to provide source code to gdb)

# Context
This is an alternative to #8538 suggested in https://github.com/NixOS/nix/pull/8538#issuecomment-1676146508

This just exposes `Store::queryValidDerivers(const StorePath & path)` in the cli.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Review by iFreilicht — CHANGES_REQUESTED** [review]
Looks pretty good! Documentation was updated in the right places and is generally useful, scope is good, and the tests mostly look good, too. Some small qualitative notes and questions, no major concerns from my side.

**Comment by iFreilicht** [comment]
Ah, one thing I forgot: You might want to add a note to `rl-next.md` so this feature shows up in the release notes :)

File: tests/dependencies.sh [hunk 1]
```diff
@@ -53,3 +53,21 @@ nix-store -q --referrers-closure "$input2OutPath" | grep "$outPath"
 # Check that the derivers are set properly.
 test $(nix-store -q --deriver "$outPath") = "$drvPath"
 nix-store -q --deriver "$input2OutPath" | grepQuiet -- "-input-2.drv"
+
+if [[ -n "${NIX_DAEMON_PACKAGE:-}" ]] && ! isDaemonNewer "2.18pre0"; then
```
[the remark this answers is not in the document]
Comment (symphorien) [reply tests/dependencies.sh:57]: Indeed I did not change the daemon, my bad, I copy pasted this from https://github.com/NixOS/nix/pull/8538

File: src/nix-store/nix-store.cc [hunk 2]
```diff
@@ -372,6 +373,21 @@ static void opQuery(Strings opFlags, Strings opArgs)
...
 
+        case qValidDerivers: {
+            StorePathSet result;
+            for (auto & i : opArgs) {
+                auto derivers = store->queryValidDerivers(store->followLinksToStorePath(i));
+                for (const auto &i: derivers) {
+                    result.insert(i);
+                }
+            }
+            auto sorted = store->topoSortPaths(result);
```
[the remark this answers is not in the document]
Comment (symphorien) [reply src/nix-store/nix-store.cc:384]: it's copy pasted from how --references &co work, tbh :)

[the remark this answers is not in the document]
Comment (iFreilicht) [reply src/nix-store/nix-store.cc:384]: Ah ok. I think it's fine then, being consistent with the behavior of other query commands makes sense.

**Review by iFreilicht — COMMENTED** [review]
Tiny suggestion for the release notes, otherwise this looks perfect now!

**Review by Ericson2314 — COMMENTED** [review]
I would like this to do CA derivations (when enabled) too

But maybe because that is an experimental feature we can do later?

**Comment by iFreilicht** [comment]
That seems to me like it should be a feature of the `queryValidDerivers` function in general, right? Doesn't it handle CA drvs already?

**Comment by roberth** [comment]
> That seems to me like it should be a feature of the `queryValidDerivers` function in general

I agree. It's about fixing a bug behind the correct interface this already uses. (and I believe that was also John's understanding of it - more of an fyi comment?)

**Review by roberth — APPROVED** [review]
This is great!

**Comment by roberth** [comment]
If you could rebase to resolve the conflict, this will be ready for merging.
Also consider https://github.com/NixOS/nix/pull/8863/files#r1305784673

**Pull request merged**